### ISSUE TITLE
How to access the confidence matrix for the instance segmentation task ?

### ISSUE THREAD

**Issue by jdolbecq** [ultralytics/ultralytics#8797]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi there!

I'm currently using your code for instance segmentation, and it's working well. During prediction, I'm able to access the binary semantic mask for each detected object in my image.

However, I'm curious if there's a way to obtain the confidence matrix that indicates, for each pixel, the probability of belonging to the class or not?

Thanks!

```python
# img is a numpy matrix of shape (n, m, 3)
result = model.predict(img)[0] 
for mask in result.masks.data:
     # mask is the binary semantic mask of a detected objet. mask is size (n, m)
     # Is it possible to get the confidence matrix instead ? It would be size (n, m, num_classes)
```

**Comment by glenn-jocher** [comment]
@jdolbecq hi there! 👋

Great to hear you're finding success with instance segmentation using our code! For accessing the confidence matrix for each pixel, you can utilize the `probs` attribute of the `Results` object returned by `model.predict()`. This attribute provides the probability scores for each class at each pixel.

Here's a quick example on how you might access it:

```python
result = model.predict(img)[0]
for prob in result.probs.data:
    # prob is the confidence matrix for a detected object, shape (n, m, num_classes)
    print(prob)
```

Please note, the structure and availability of this data might vary based on the specific model and version you're using. If `probs` doesn't directly give what you need, you might need to adjust based on the model's output documentation.

Hope this helps! Let us know if you have any more questions. 😊

**Comment by jdolbecq** [comment]
Thank you for your quick answer !

I'm confused because the 'result.probs' command returns me 'None'. 

I have trained a model like this
 ```python
model = YOLO('yolov8n-seg.pt')
results = model.train(data=my_yaml_file, imgsz=1024)
 ```

And I do a prediction like this

 ```python
model = YOLO("path/to/my/weights/file.pt")
results = model.predict(img, imgsz=1024, max_det=-1) # img is a numpy array of shape (1024, 1024, 3)
result = results[0]
result.prob # is None
result.masks.data # tensor of size (663, 1024, 1024])
 ```

Is it necessary to activate an argument to obtain the probs value? Or is this simply not possible with the model I'm using?

**Comment by glenn-jocher** [comment]
Hi @jdolbecq! 😊

It looks like there's a bit of confusion regarding accessing the confidence scores for each pixel, especially since `result.probs` is returning `None` for you. For instance segmentation tasks using YOLOv8 models, the primary outputs are masks, and the probability scores are typically associated with the detection boxes rather than individual pixels within the masks.

If you're looking to get confidence scores for each detected object (which the masks correspond to), you should look at the `results.boxes` attribute, which contains the bounding boxes along with their confidence scores and class IDs. Here's how you can access them:

```python
model = YOLO("path/to/my/weights/file.pt")
results = model.predict(img, imgsz=1024)  # Assuming img is your input image
for box in results[0].boxes:
    print(f"Confidence: {box.conf}, Class ID: {box.cls}")
```

This will give you the confidence score for each detected object. Unfortunately, getting a per-pixel confidence score directly from the model's output isn't standard for segmentation tasks in YOLOv8 as the model focuses on providing binary masks indicating the presence of a class within each pixel.

If you're specifically looking for pixel-wise class probabilities, this might require a custom implementation or a different model architecture designed to output such information.

I hope this clarifies things a bit! Let me know if you have any more questions.

**Comment by glenn-jocher** [comment]
@jdolbecq hi there! 😊

You're right, accessing a per-pixel confidence matrix isn't a standard output for YOLOv8 models, but your interest in diving deeper into the code is commendable! The transition from confidence scores to binary masks typically involves thresholding the model's output probabilities to decide whether a pixel belongs to a class.

For YOLOv8, this operation is closely tied to the model's architecture and post-processing steps. While I can't point you to a specific line of code without knowing the exact version of YOLOv8 you're using, I'd recommend looking into the `forward` method of the model where the inference logic is defined. This is usually where the raw model outputs are processed into the final predictions.

Implementing a custom architecture or modifying the existing one for pixel-wise confidence scores would involve adjusting the model's final layers to output class probabilities for each pixel without applying a threshold. You might also need to modify the post-processing steps to handle these probabilities appropriately.

Here's a very simplified example of what modifying the output layer could look like, assuming you're working with PyTorch:

```python
import torch.nn as nn

# Example modification of the last layer to output class probabilities
model.last_layer = nn.Sequential(
    nn.Conv2d(in_channels=model.last_layer_in_channels, out_channels=num_classes, kernel_size=1),
    nn.Sigmoid()  # Assuming binary classification per pixel
)
```

Remember, this is a simplified example and actual implementation details will vary based on your specific requirements and the YOLOv8 model structure.

I hope this gives you a starting point! Your exploration could indeed be very useful for the community. If you make progress, consider sharing your findings or contributing back. Good luck! 🚀

**Comment by mluerig** [comment]
just wanted to follow up on this, but I need some more help - I'm not sure how to access the last layer (keep getting "not subscriptable" error). I'm using the [11x-seg model](https://github.com/ultralytics/assets/releases/download/v8.3.0/yolo11x-seg.pt)

**Comment by jdolbecq** [comment]
Hi @mluerig !

I managed to access the confidence maps by modifying the _process_mask_ and _process_mask_native_ functions in the _ultralytics/ultralytics/utils/ops.py_ file. In these two functions, the line` return masks.gt_(0.0) ` must be replaced by `return masks`. In this way, `result.masks.data` returns the confidence mask and not the binary mask.

**Comment by mluerig** [comment]
@jdolbecq can confirm this works - I wrote a monkeypatch that can turn this on and off.

@glenn-jocher since this is such a simple thing to do, why not make it an argument? or even simpler, default to return confidence arrays instead of binary masks - I think the current functionality and logic of the results object doesn't have to be changed at all for this

**Comment by glenn-jocher** [comment]
Hi @mluerig and @jdolbecq, thank you for sharing your insights! We appreciate community contributions and will consider this enhancement for future releases. For those interested in this functionality, we recommend monitoring our GitHub for updates or opening a Feature Request issue to discuss implementation details. The current design prioritizes efficiency for common use cases, but we're always evaluating improvements. 🚀

**Comment by raficabral** [comment]
@mluerig @jdolbecq Thank you for your support, I am facing the same issue. I tried to replace `return masks.gt_(0.0)` to `return masks`. Could you say if you changed anything else apart from process_mask and process_mask_native? My return is empty.

**Comment by mluerig** [comment]
this was pretty much all I did:
    
    import torch
    import warnings
    
    
    try:
        import ultralytics.utils.ops as ops
        original_process_mask = ops.process_mask
        original_process_mask_native = ops.process_mask_native
    except ImportError:
        warnings.warn("Failed to import Ultralytics. Some functionalities may not work.", ImportWarning)
    
    
    #%% monkey-patch
    
    
    
    # Define modified functions (returning confidence masks instead of binary masks)
    def modified_process_mask(protos, masks_in, bboxes, shape, upsample=False):
        c, mh, mw = protos.shape  # CHW
        ih, iw = shape
        masks = (masks_in @ protos.float().view(c, -1)).view(-1, mh, mw)  # CHW
        width_ratio = mw / iw
        height_ratio = mh / ih
    
        downsampled_bboxes = bboxes.clone()
        downsampled_bboxes[:, 0] *= width_ratio
        downsampled_bboxes[:, 2] *= width_ratio
        downsampled_bboxes[:, 3] *= height_ratio
        downsampled_bboxes[:, 1] *= height_ratio
    
        # masks = ops.crop_mask(masks, downsampled_bboxes)  # CHW
        if upsample:
            masks = torch.nn.functional.interpolate(masks[None], shape, mode="bilinear", align_corners=False)[0]  # CHW
        return torch.sigmoid(masks)  # Normalize output to be in [0,1]
    
    def modified_process_mask_native(protos, masks_in, bboxes, shape):
        c, mh, mw = protos.shape  # CHW
        masks = (masks_in @ protos.float().view(c, -1)).view(-1, mh, mw)
        masks = ops.scale_masks(masks[None], shape)[0]  # CHW
        # masks = ops.crop_mask(masks, bboxes)  # CHW
        return torch.sigmoid(masks)  # Normalize output to be in [0,1]
    
    # Function to toggle between original and modified functions
    def toggle_mask_conf(use_modified=True):
        if use_modified:
            ops.process_mask = modified_process_mask
            ops.process_mask_native = modified_process_mask_native
        else:
            ops.process_mask = original_process_mask
            ops.process_mask_native = original_process_mask_native

**Comment by glenn-jocher** [comment]
Thank you @mluerig for sharing this detailed implementation! This is really helpful for the community.

@raficabral, the empty returns you're experiencing might be because the monkey patch needs more than just replacing the return statement. As you can see from @mluerig's code, they're:

1. Using `torch.sigmoid(masks)` to normalize the outputs to [0,1] range
2. Commenting out the `crop_mask` operations 
3. Creating a toggle function to switch between behaviors

The key difference appears to be applying the sigmoid activation to the raw mask values before returning them, which ensures they're properly scaled confidence values. Without this step, you might get values outside a meaningful range.

We'll consider adding this as a configurable option in future releases. In the meantime, this monkey-patching approach is a good workaround. If you're still having issues after implementing these changes, please share more details about your specific setup.
